sky/provision/instance_setup.py

```diff
@@ -114,8 +115,38 @@ def _parallel_ssh_with_cache(func, cluster_name: str, stage_name: str,
             results.append(
                 pool.submit(wrapper(func), runner, metadata, log_path_abs))
 
-        for future in results:
-            future.result()
+        return [future.result() for future in results]
+
+
+@_log_start_end
+def initialize_docker(cluster_name: str, docker_config: Dict[str, Any],
+                      cluster_metadata: common.ClusterMetadata,
+                      ssh_credentials: Dict[str, str]) -> Optional[str]:
+    """Setup docker on the cluster."""
+    if not docker_config:
+        return None
+    _hint_worker_log_path(cluster_name, cluster_metadata, 'initialize_docker')
+
+    @_auto_retry
+    def _initialize_docker(runner: command_runner.SSHCommandRunner,
+                           metadata: common.InstanceMetadata, log_path: str):
+        del metadata  # Unused.
+        docker_user = docker_utils.DockerInitializer(docker_config, runner,
+                                                     log_path).initialize()
+        logger.debug(f'Initialized docker user: {docker_user}')
+        return docker_user
+
+    docker_users = _parallel_ssh_with_cache(
+        _initialize_docker,
+        cluster_name,
+        stage_name='initialize_docker',
+        # Should not cache docker setup, as it needs to be
+        # run every time a cluster is restarted.
+        digest=str(time.time()),
+        cluster_metadata=cluster_metadata,
+        ssh_credentials=ssh_credentials)
+    logger.debug(f'All docker users: {docker_users}')
```

Comment:
nit: assert every entry in the list is the same here?

Reply:
Good point! Added.